Add a script to report outdated externalized dependencies.

tomalec opened the pull request:
This is a Request For Comments, on a potential DevX improvement.

A script that reports outdated dependencies, but **only** those dependencies which were **externalized** with Dependency Extraction Webpack Plugin.
So the dependencies that are beyond the control of our bundle, and may actually run on a different (newer) version than we install locally.

### Changes proposed in this Pull Request:

- Report externalized dependencies to `/.externalized.json`.
- Add a script to report outdated externalized dependencies.
	The idea is to help to inspect how wide is the range of granular dependencies we need to anticipate.
- Add a script to report outdated bundled dependencies,
	So we can check and update those dependencies that do not require and WP&WC compatibility hassle.

:warning:  So far this PR uses `trunk` version of `@wordpress/dependency-extraction-webpack-plugin` to be able to use unreleased
https://github.com/WordPress/gutenberg/pull/35106, hopefully, we'd be able to use released version soon.


### Screenshots:
![image](https://user-images.githubusercontent.com/17435/144277662-66b7adb3-34aa-4d01-98e4-6bce3161c3c5.png)




### Detailed test instructions:

0. (optional) `npm i && npm run dev `
1. `npm run outdated:dewp`
2. `npm run outdated:nondewp`

### Additional notes
1. I was thinking of a better place for the `.externalized.json` file not to bloat the root, but had no idea. We may also add it to `.gitignore`, but then we would need to always run `npm run dev` before `dewp:outdated` what actually makes some sense, to be sure to get always up to date report.
2. `outdated:dewp` script gives the information of the latest version of granular packages, which may or may not be present in the latest version of WordPress, Gutenberg, or WooCommerce. Also, quite frequently we'd need also info about the oldest version of granular packages used by the oldest version of WP/WC we support. But that's something that would rather employ  `wp-scripts` https://github.com/WordPress/gutenberg/pull/36005#
3. The above means the outcome of this script should not make us update those dependencies, as the minimal WP version may still run on the legacy version we have locally, to run tests against. This script rather gives us the information, how far the current legacy versions drag us down and provide a reason to consider updating the minimal WP/WC version. Or to double-check that the newer versions didn't introduce any backward-incompatible changes.
- [x] Needs to be updated after/if https://github.com/woocommerce/google-listings-and-ads/pull/1128 is merged.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

tomalec committed:
Add a script to report outdated externalized dependencies.

Report externa;ized dependencies to `/.externalized.json`.
Help to inspect how wide is the range of granular dependencies we need to anticipate.

Use `trunk` version of `@wordpress/dependency-extraction-webpack-plugin` to be able to use unreleased
https://github.com/WordPress/gutenberg/pull/35106

eason9487 commented on package-lock.json:
```diff
@@ -6722,13 +6722,30 @@
-			"version": "2.9.0",
-			"resolved": "https://registry.npmjs.org/@wordpress/dependency-extraction-webpack-plugin/-/dependency-extraction-webpack-plugin-2.9.0.tgz",
-			"integrity": "sha512-Eo8ByPd3iZ6az4UmdLD2xYLp1/7os/H80l28Y5OlS4DozkD3vcWCBReynWoBax74u3oJ9wWN5b/8oSxGwIKXYQ==",
+			"version": "https://gitpkg.now.sh/WordPress/gutenberg/packages/dependency-extraction-webpack-plugin?trunk",
```
💅 
Since the `trunk` of gutenberg is continuously updating, checksum failure occurs when running `npm install`, it'd be easier to test PR with the specified commit hash.

```suggestion
			"version": "https://gitpkg.now.sh/WordPress/gutenberg/packages/dependency-extraction-webpack-plugin?6d87c733c77c33f821fde94e5a8203212aec422b",
```

eason9487 commented on package.json:
```diff
@@ -49,7 +50,7 @@
 		"@woocommerce/e2e-utils": "^0.1.6",
 		"@woocommerce/eslint-plugin": "^1.2.0",
-		"@wordpress/dependency-extraction-webpack-plugin": "^2.7.0",
+		"@wordpress/dependency-extraction-webpack-plugin": "https://gitpkg.now.sh/WordPress/gutenberg/packages/dependency-extraction-webpack-plugin?trunk",
```
💅  Same as above.

```suggestion
		"@wordpress/dependency-extraction-webpack-plugin": "https://gitpkg.now.sh/WordPress/gutenberg/packages/dependency-extraction-webpack-plugin?6d87c733c77c33f821fde94e5a8203212aec422b",
```

eason9487 commented:
> 1. I was thinking of a better place for the `.externalized.json` file not to bloat the root, but had no idea. We may also add it to `.gitignore`, but then we would need to always run `npm run dev` before `dewp:outdated` what actually makes some sense, to be sure to get always up to date report.

What about adding a pre script as below for `dewp:outdated`, then we could add `.externalized.json` to *.gitignore*? Or maybe just merge the pre script to the front of `dewp:outdated`.

```json
"predewp:outdated": "npm run dev -- --display=errors-only",
```

tomalec commented:
I start to think, maybe it's worth keeping that file in git. That way we could, in the future:
- notice explicitly, if a PR adds/removes something to/from the list, so the review may require additional caution,
- run some automation on top of that, like GH dependabot,
- do not have to `npm run build` just after `git clone` to see the list of dependencies.
- inspect easily in web the list of externalized dependencies
	- so we could reference it in GH discussions
	- that's something I'd like to see for WooCommerce - which of its dependencies are bundled and at versions specified in `package.json` which are not.

tomalec committed:
Install specific commit from DEWP

to avoid checksum failures.

Co-authored-by: Eason <eason.su.tw@gmail.com>

tomalec committed:
Add a npm script to list outdated non-externalized deps

tomalec committed:
Rename npm script `dewp:outdated` to `outdated:dewp`

tomalec commented:
I added another script, to list also non externalized dependencies. This may be useful when we just want to bump "local"=bundled dependencies.

tomalec committed:
Account for colored output from `npm outdated`.

in `outdated:nondewp`.

tomalec committed:
Merge branch 'develop' into add/dewp-outdated

tomalec commented:
Also, if we found ourselves using this script often, and useful for other repos, we may try to PR it to `wp-scripts`.

tomalec committed:
Merge branch 'develop' into add/dewp-outdated

tomalec committed:
Update package-lock.json

puntope commented:
> ### Detailed test instructions:
> 0. (optional) `npm run dev `
> 1. `npm run outdated:dewp`
> 2. `npm run outdated:nondewp`

Step -1: `npm i` otherwise I got an error on `npm run dev`

puntope commented:
❓ 

I take as granted that the `Wanted` column is the version we should upgrade in package.json?

<img width="1151" alt="Screenshot 2021-12-10 at 10 26 20" src="https://user-images.githubusercontent.com/5908855/145551079-18101f6e-0a96-4c8d-a91e-4e6566bf45f0.png">

puntope commented:
❓

I'm unsure about the responsibility of `externalized.json` what should we do with the information provided on it?

puntope commented on package.json:
```diff
@@ -49,7 +51,7 @@
 		"@woocommerce/e2e-environment": "^0.2.3",
 		"@woocommerce/e2e-utils": "^0.1.6",
 		"@woocommerce/eslint-plugin": "^1.2.0",
-		"@wordpress/dependency-extraction-webpack-plugin": "^2.7.0",
```
What is the advantage again to have it remote instead of using standard  `^2.7.0` versioning?

ecgan replied:
@puntope , it is related to the following, which is mentioned in the PR description.

> :warning:  So far this PR uses `trunk` version of `@wordpress/dependency-extraction-webpack-plugin` to be able to use unreleased https://github.com/WordPress/gutenberg/pull/35106, hopefully, we'd be able to use released version soon.

eason9487 commented:
> I start to think, maybe it's worth keeping that file in git. That way we could, in the future:
> 
> * notice explicitly, if a PR adds/removes something to/from the list, so the review may require additional caution,
> * [...]

I think these are good possible uses. But the order of packages output is not stable in `.externalized.json`, which may cause false alarms to occur more frequently than the actual package changes. And it may also come up sometimes when `git add/diff` and be a bit distracting. Probably need to have a postscript or some handler to normalize it for the first idea.

tomalec commented:
> I think these are good possible uses. But the order of packages output is not stable in .externalized.json, which may cause false alarms to occur more frequently than the actual package changes.

Great catch! I'd make a PR to sort them by DEWP. https://github.com/WordPress/gutenberg/pull/37377

tomalec commented:
@puntope 
> I'm unsure about the responsibility of `externalized.json` what should we do with the information provided on it?

This gives us the information: 
How the current version of `@wordpress/-`, `@woocommerce/dependency-extraction-webpack-plugin`, our `webpack.config.js` and all the `import` statements was eventually resolved. Which packages from `package.json.dependencies` were externalized from the bundle = are not included in our bundle, and may differ in runtime from what we have in `node_modules` and `package.json`.

tomalec committed:
Merge branch 'develop' into add/dewp-outdated

tomalec committed:
Install `next` version of `@wordpress/dependency-extraction-webpack-plugin`

from official repo.

tomalec committed:
Update `.externalized.json`

with sorted dependencies, after re-run `npm run build` with the latest version of DEWP.

tomalec commented:
@eason9487 @puntope The dependencies are now ordered, and we install DEWP from the official source.

WDYT of merging it?

eason9487 reviewed (approved):
Thanks for the quickly pushing the patch to upstream for getting stable result of *.externalized.json*. I think it's good to merge! 💯
